File: api/src/main/java/net/techcable/sonarpet/utils/compat/GuavaVersion.java
package net.techcable.sonarpet.utils.compat;

import java.util.concurrent.Executor;
import java.util.stream.Collector;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;

public interface GuavaVersion {
    <T> Collector<T, ?, ImmutableList<T>> immutableListCollector();
    <T> Collector<T, ?, ImmutableSet<T>> immutableSetCollector();
    Executor directExecutor();

    GuavaVersion INSTANCE = acquireInstance();
    static GuavaVersion acquireInstance() {
        if (INSTANCE != null) return INSTANCE;
        try {
            ImmutableList.class.getMethod("toImmutableList");
            return new ModernGuava();
        } catch (NoSuchMethodException e) {
            return new AncientGuava();
        }
    }
}

File: api/src/main/kotlin/net/techcable/sonarpet/utils/compat/GuavaVersion.kt
package net.techcable.sonarpet.utils.compat

import com.google.common.collect.ImmutableList
import java.util.concurrent.Executor
import java.util.stream.Collector

interface GuavaVersion {
    fun <E> immutableListCollector(): Collector<E, *, ImmutableList<E>>
    fun directExecutor(): Executor

    companion object {
        val DETECTED_VERSION = detectVersion()
        private fun detectVersion(): GuavaVersion {
            try {
                ImmutableList::class.java.getMethod("toImmutableList")
                return ModernGuava
            } catch (e: NoSuchMethodException) {
                return AncientGuava
            }
        }
    }
}